bin/server.py

```diff
@@ -37,20 +38,24 @@ class MockRootBackend(RootBackend):
     def __init__(self, orders: MockOrderDB) -> None:
         self._orders: MockOrderDB = orders
 
-    async def get_orders(self, request: Request) -> OrderCollection:
+    async def get_orders(
+        self, request: Request
+    ) -> Result[OrderCollection, Maybe[Exception]]:
         """
         Show all orders.
         """
-        return OrderCollection(features=list(self._orders.values()))
+        return Success(OrderCollection(features=list(self._orders.values())))
 
-    async def get_order(self, order_id: str, request: Request) -> Order:
+    async def get_order(
```

Comment:
I got some type errors when I opened this code in my editor. Not specifically here, but as I looked more at it I started wondering if this was correct.

Why is this function not like this:

```python
    async def get_order(
        self, order_id: str, request: Request
    ) -> Result[Maybe[Order], Exception]:
        """
        Show details for order with `order_id`.
        """
        try:
            return Success(self._orders.get(order_id))
        except Exception as e:
            return Failure(e)
```

I don't think it would change the `match` handling the return values beyond the case for no order found being `Success(None)` instead of `Failure(None)`.

Reply:
Also, would it make sense to use the `@safe` decorator here (and on the other functions) to simplify by removing the `try/except` and other `returns`-related bits?

Reply:
Ha, my suggestion has a type error. It's the maybes, that's the problem. No where here do we actually have a `Maybe` type value.

Now, if our `.get()` method returned a `Maybe` we'd be good. But that's a built-in dict method, so it obviously doesn't. It looks like the way to fix that is to wrap that call with `Maybe.from_optional()`:

```python
    async def get_order(
        self, order_id: str, request: Request
    ) -> Result[Maybe[Order], Exception]:
        """
        Show details for order with `order_id`.
        """
        try:
            return Success(Maybe.from_optional(self._orders.get(order_id)))
        except Exception as e:
            return Failure(e)
```

Or, with decorators maybe (no pun intended) this could be like:

```python
    @safe
    @maybe
    async def get_order(
        self, order_id: str, request: Request
    ) -> Result[Maybe[Order], Exception]:
        """
        Show details for order with `order_id`.
        """
        return self._orders.get(order_id)
```

?

Reply:
`Failure(e)` should be `Failure(Some(e))` -- I've fixed those, though I'm not sure why pre-commit mypy or my ide didn't find that.

Reply:
I think 

```
if order := self._orders.get(order_id):
            return Success(order)

        return Failure(Nothing)
```

is pretty clear code as it is, and those two suggestions make it harder to understand. The first one is more complicated with a try/except block and 4 nested expressions, and the second is more complicated because you have to look at both the annotations and actual code. I think for something that's a test/example that the existing code is as clear as it  can be.